x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/ReservedRoleMappingAction.java

```diff
@@ -0,0 +1,145 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.action.rolemapping;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.GroupedActionListener;
+import org.elasticsearch.reservedstate.NonStateTransformResult;
+import org.elasticsearch.reservedstate.ReservedClusterStateHandler;
+import org.elasticsearch.reservedstate.TransformState;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xpack.core.security.action.rolemapping.DeleteRoleMappingRequest;
+import org.elasticsearch.xpack.core.security.action.rolemapping.PutRoleMappingRequest;
+import org.elasticsearch.xpack.core.security.authc.support.mapper.ExpressionRoleMapping;
+import org.elasticsearch.xpack.security.authc.support.mapper.NativeRoleMappingStore;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.common.xcontent.XContentHelper.mapToXContentParser;
+
+/**
+ * This Action is the reserved state save version of RestPutRoleMappingAction/RestDeleteRoleMappingAction
+ * <p>
+ * It is used by the ReservedClusterStateService to add/update or remove role mappings. Typical usage
+ * for this action is in the context of file based settings.
+ */
+public class ReservedRoleMappingAction implements ReservedClusterStateHandler<List<ExpressionRoleMapping>> {
+    public static final String NAME = "role_mappings";
+
+    private final NativeRoleMappingStore roleMappingStore;
+
+    /**
+     * Creates a ReservedRoleMappingAction
+     *
+     * @param roleMappingStore requires {@link NativeRoleMappingStore} for storing/deleting the mappings
+     */
+    public ReservedRoleMappingAction(NativeRoleMappingStore roleMappingStore) {
+        this.roleMappingStore = roleMappingStore;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    private Collection<PutRoleMappingRequest> prepare(List<ExpressionRoleMapping> roleMappings) {
+        List<PutRoleMappingRequest> requests = roleMappings.stream().map(rm -> PutRoleMappingRequest.fromMapping(rm)).toList();
+
+        var exceptions = new ArrayList<String>();
+        for (var request : requests) {
+            // File based defined role mappings are allowed to use MetadataUtils.RESERVED_PREFIX
+            var exception = request.validate(false);
+            if (exception != null) {
+                exceptions.add(exception.getMessage());
+            }
+        }
+
+        if (exceptions.isEmpty() == false) {
+            throw new IllegalStateException(String.join(", ", exceptions));
```

Comment:
I'd go with
```suggestion
            var illegalStateException = new IllegalStateException("error on validating put role mapping requests");
            exceptions.forEach(illegalStateException::addSuppressed);
```

I am a bit surprised about the choice of IllegalStateException type. Errors here are caused by users which is more of IllegalArgument. But I notice this error type is already used elsewhere. So I am also OK with it.